core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -249,7 +249,14 @@ private[spark] class ExecutorAllocationManager(
    * yarn-client mode when AM re-registers after a failure.
    */
   def reset(): Unit = synchronized {
-    initializing = true
+    /**
+     * When some tasks need to be scheduled and initial executor = 0, resetting the initializing
+     * field may cause it to not be set to false in yarn.
+     * SPARK-20079: https://issues.apache.org/jira/browse/SPARK-20079
+     */
+    if (maxNumExecutorsNeeded() == 0) {
+      initializing = true
```

Comment:
I think the purpose of "initializing" is to avoid unnecessary executors ramp down before the stage submitted or executor timeout. For example if min executor number is 0, initial number is 10. If "initializing" is set to false, executor number will ramp down to 0 immediately, and during this time if stage is submitted, then still requires unnecessary executor ramp up to meet this stage's requirement.

In the AM restart scenario, if we set "initializing" to false in `reset`, then we will may also meet the situation mentioned above, I think that's possible.